### PR TITLE
Recobundlex standalone

### PR DESCRIPTION
**Pull request by frheault**
Transfert of the single-atlas single-parameters Recobundles from scilpy as well as multi-atlas multi-parameters Recobundles from my personal repository (and all updated and tested in Python3)

One PR for both makes it easier to test, compare and since the algorithms and parameters are very similar in both it makes the review easier and faster.

Here is something to test it with, and a bash script to help.
https://drive.google.com/file/d/16Nr-udxbAhyLCE5x6qO5V8B-OGt37n2C/view?usp=sharing

### CODE REVIEW

**Comment by mdesco** [comment]
We should add Mario Ocopamdo to test and review also. He's been your biggest user in the past 2 months.

**Comment by mdesco** [comment]
Finished testing. All works well! Amazing stuff, especially the recognize_multi_bundles. Good job!

**Comment by mdesco** [comment]
I approve

File: scilpy/segment/voting_scheme.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+        seeds : list
+            List of seed for the RandomState
+        """
+
+        # Load the subject tractogram
+        timer = time()
+        tractogram = nib.streamlines.load(input_tractogram_path)
+        wb_streamlines = tractogram.streamlines
+        logging.debug('Tractogram %s with %s streamlines ' +
+                      'is loaded in %s', input_tractogram_path,
```
Comment (arnaudbore) [inline scilpy/segment/voting_scheme.py]: add seconds after `is loaded in`

File: scripts/scil_recognize_multi_bundles.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                   help='Path of the config file (json)')
+    p.add_argument('models_directories', nargs='+',
+                   help='Path for the directories containing model.')
+    p.add_argument('transformation',
+                   help='Path for the transformation to model space.')
+
+    p.add_argument('--output_dir', default='voting_results/',
+                   help='Path for the output directory.')
+    p.add_argument('--log_level', default='INFO',
+                   choices=['DEBUG', 'INFO', 'WARNING', 'ERROR', 'CRITICAL'],
```
Comment (arnaudbore) [inline scripts/scil_recognize_multi_bundles.py]: CRITICAL is not used we should remove it.

File: scripts/scil_recognize_multi_bundles.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+import coloredlogs
+import numpy as np
+
+from scilpy.io.utils import add_overwrite_arg, assert_inputs_exist
+from scilpy.segment.voting_scheme import VotingScheme
+
+
+def _buildArgsParser():
+    p = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter,
+                                description=__doc__)
```
Comment (GuillaumeTh) [inline scripts/scil_recognize_multi_bundles.py]: Can you add the recobundle reference in the parser ? e.g. `epilog="References: [1] Girard, G., Whittingstall K., Deriche, R., "
               "and Descoteaux, M. (2014). Towards quantitative connectivity "
               "analysis: reducing tractography biases. Neuroimage, 98, "
               "266-278."`

File: scripts/scil_recognize_single_bundle.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import numpy as np
+
+from scilpy.io.utils import (add_overwrite_arg,
+                             assert_inputs_exist,
+                             assert_outputs_exists)
+
+
+def buildArgsParser():
+    p = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter,
+                                description=__doc__)
```
Comment (GuillaumeTh) [inline scripts/scil_recognize_single_bundle.py]: Can you add the recobundle reference in the parser ? e.g. `epilog="References: [1] Girard, G., Whittingstall K., Deriche, R., " "and Descoteaux, M. (2014). Towards quantitative connectivity " "analysis: reducing tractography biases. Neuroimage, 98, " "266-278."`

File: scilpy/segment/recobundlesx.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+        clusters : list
+            Streamlines that were recognized by Recobundles and these
+            parameters
+        """
+        self._cluster_model_bundle(model_bundle, model_clust_thr,
+                                   identifier=identifier)
+
+        if not self._reduce_search_space():
+            if identifier:
+                logging.error('%s did not find any neighbors in ' +
```
Comment (GuillaumeTh) [inline scilpy/segment/recobundlesx.py]: Can you use .format instead of `%s`

File: scilpy/segment/recobundlesx.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+        self.pruned_indices_per_clusters = self._prune_what_not_in_model(
+            transf_streamlines,
+            bundle_pruning_thr=bundle_pruning_thr)
+
+        return self.pruned_streamlines
+
+    def _cluster_model_bundle(self, model, model_clust_thr, identifier=None):
+        """
+        Wrapper function to compute QBx for the model and logging informations
```
Comment (GuillaumeTh) [inline scilpy/segment/recobundlesx.py]: Can you add documentation about parameters.

File: scilpy/segment/recobundlesx.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        """
+        thresholds = [30, 20, 15, model_clust_thr]
+        self.model_cluster_map = qbx_and_merge(model, thresholds,
+                                               nb_pts=self.nb_points,
+                                               rng=self.rng,
+                                               verbose=False)
+        self.model_centroids = self.model_cluster_map.centroids
+        len_centroids = len(self.model_centroids)
+        if len_centroids > 1000:
+            logging.warning('Model %s simplified at threshod ' +
```
Comment (GuillaumeTh) [inline scilpy/segment/recobundlesx.py]: Same thing here, can you use .format please.

File: scilpy/segment/recobundlesx.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+                            '%smm with %s centroids', identifier,
+                            str(model_clust_thr),
+                            str(len_centroids))
+
+    def _reduce_search_space(self, neighbors_reduction_thr=18):
+        """
+        Wrapper function to discard clusters from the tractogram too far from
```
Comment (GuillaumeTh) [inline scilpy/segment/recobundlesx.py]: Parameters documentation please.

File: scilpy/segment/recobundlesx.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+                                               bounds=bounds_dof[:9],
+                                               num_threads=slr_num_thread)
+            slm = slr.optimize(static, moving)
+
+        return transform_streamlines(self.neighb_streamlines, slm.matrix)
+
+    def _prune_what_not_in_model(self, neighbors_to_prune, bundle_pruning_thr=10,
+                                 neighbors_cluster_thr=8):
+        """
+        Wrapper function to prune clusters from the tractogram too far from
```
Comment (GuillaumeTh) [inline scilpy/segment/recobundlesx.py]: Parameters documentation please.

File: scilpy/segment/voting_scheme.py [hunk 10]
```diff
@@ -0,0 +1,435 @@
...
+        for i in range(len(bundles_filepath)):
+            missing_count = 0
+            missing_files = []
+            for j in range(len(bundles_filepath[i])):
+                if not os.path.isfile(bundles_filepath[i][j]):
+                    missing_count += 1
+                    missing_files.append(bundles_filepath[i][j])
+
+            if missing_count == len(bundles_filepath[i]):
+                logging.warning('None of the %s exist, this bundle' +
```
Comment (GuillaumeTh) [inline scilpy/segment/voting_scheme.py]: Same thing here and below, can you use .format ?

File: scilpy/segment/voting_scheme.py [hunk 11]
```diff
@@ -0,0 +1,435 @@
...
+                     len(self.atlas_dir),
+                     len(bundle_names_exist))
+        logging.debug('The models use for RecobundlesX ' +
+                      'will be %s', bundles_filepath_exist)
+
+        return bundle_names_exist, bundles_filepath_exist
+
+    def _load_bundles_dictionary(self, bundles_filepath):
+        """
+        Load all model bundles and store them in a dictionnary where the
```
Comment (GuillaumeTh) [inline scilpy/segment/voting_scheme.py]: A small documentation for the parameters please.

File: scilpy/segment/voting_scheme.py [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+                          len(bundle))
+            if len(bundle) > 5000:
+                logging.warning('%s has above 5000 streamlines', filename)
+
+        return model_bundles_dict
+
+    def _find_max_in_sparse_matrix(self, bundle_id, min_vote,
+                                   streamlines_wise_vote, bundles_wise_vote):
+        """
+        Will find the maximum values of a specific row (bundle_id), make
```
Comment (GuillaumeTh) [inline scilpy/segment/voting_scheme.py]: parameters documentation please

File: scilpy/segment/voting_scheme.py [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+        """
+        results_dict = {}
+        for bundle_id in range(len(bundle_names)):
+            streamlines_id = self._find_max_in_sparse_matrix(bundle_id,
+                                                             minimum_vote,
+                                                             streamlines_wise_vote,
+                                                             bundles_wise_vote)
+
+            if not streamlines_id.size:
+                logging.error('%s final recognition got %s streamlines',
```
Comment (GuillaumeTh) [inline scilpy/segment/voting_scheme.py]: Same here and below, please replace with .format

File: scilpy/segment/voting_scheme.py [hunk 14]
```diff
@@ -0,0 +1,435 @@
...
+            Number of processes used for the parallel bundle recognition
+        seeds : list
+            List of seed for the RandomState
+        """
+
+        # Load the subject tractogram
+        timer = time()
+        tractogram = nib.streamlines.load(input_tractogram_path)
+        wb_streamlines = tractogram.streamlines
+        logging.debug('Tractogram %s with %s streamlines ' +
```
Comment (GuillaumeTh) [inline scilpy/segment/voting_scheme.py]: Same here and below, please replace with .format

File: scilpy/segment/voting_scheme.py [hunk 15]
```diff
@@ -0,0 +1,435 @@
...
+
+        # Cluster the whole tractogram only once per possible clustering threshold
+        rbx_all = {}
+        base_thresholds = [45, 35, 25]
+        for seed in seeds:
+            rng = np.random.RandomState(seed)
+            for clustering_thr in tractogram_clustering_thr:
+                timer = time()
+                # If necessary, add an extra layer (more optimal)
+                if clustering_thr < 15:
```
Comment (GuillaumeTh) [inline scilpy/segment/voting_scheme.py]: Why 15 ? I don't understand

File: scripts/scil_recognize_multi_bundles.py [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+
+    p.add_argument('--multi_parameters', type=int, default=1,
+                   help='Pick parameters from the potential combinations\n' +
+                   'Will multiply the number of time Recobundles is ran.')
+    p.add_argument('--minimal_vote_ratio', type=float, default=0.5,
+                   help='Streamlines will only be considered for saving if\n ' +
+                   'recognized often enough.')
+
+    p.add_argument('--tractogram_clustering_thr',
+                   type=int, default=[12], nargs='+',
```
Comment (GuillaumeTh) [inline scripts/scil_recognize_multi_bundles.py]: Why [12] and not 12 ?

File: scripts/scil_recognize_multi_bundles.py [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+                   help='Input tractogram clustering thresholds ' +
+                   '[%(default)smm].')
+
+    p.add_argument('--processes', type=int, default=1,
+                   help='Number of thread used for computation [%(default)s].')
+    p.add_argument('--seeds', type=int, default=[None], nargs='+',
```
Comment (GuillaumeTh) [inline scripts/scil_recognize_multi_bundles.py]: Why [None] and not None

File: scripts/scil_recognize_multi_bundles.py [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+                   help='Random number generator seed [%(default)s]\n' +
+                   'Will multiply the number of time Recobundles is ran.')
+    p.add_argument('--inverse', action='store_true',
+                   help='Use the inverse transformation.')
```
Comment (GuillaumeTh) [inline scripts/scil_recognize_multi_bundles.py]: Maybe add this option must be used if the transform come from ANTs

File: scripts/scil_recognize_single_bundle.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    p.add_argument('--pruning_thr', type=float, default=6,
+                   help='MDF threshold used for final streamlines selection '
+                   '[%(default)smm].')
+
+    p.add_argument('--slr_threads', type=int, default=None,
+                   help='Number of threads for SLR [all].')
+    p.add_argument('--seed', type=int, default=None,
+                   help='Random number generator seed [%(default)s].')
+    p.add_argument('--inverse', action='store_true',
+                   help='Use the inverse transformation.')
```
Comment (GuillaumeTh) [inline scripts/scil_recognize_single_bundle.py]: Maybe add this option must be used if the transform come from ANTs

Comment (frheault) [reply scripts/scil_recognize_single_bundle.py]: Well, this can be as confusing. It also depend on how you called ANTs.
Depending on the moving/target option used with ANTs, maybe you don't need it. (e.g A->B could require --inverse, but not B->A)

Comment (GuillaumeTh) [reply scripts/scil_recognize_single_bundle.py]: I agree. forget this comment

File: scripts/scil_recognize_multi_bundles.py (hunk 17 above)
Comment (frheault) [reply scripts/scil_recognize_multi_bundles.py]: It is an nargs, so it saves a if later on. Instead of checking if it is None and then I would have to replace it with [None]

File: scilpy/segment/voting_scheme.py (hunk 15 above)
Comment (frheault) [reply scilpy/segment/voting_scheme.py]: It is because QBx requires threshold that are decreasing, but also having a useless layer is suboptimal. Instead of systematically adding a new layer or having corner case, there is a check. 

Imagine you have a clustering threshold of **8**: [45, 35, 25, 8] (big gap, suboptimal)
But I can't systematically add a layer at 15 in case some use a clustering threshold of **18**: [45, 35, 25, 15, 18] (Inversion)

So basically the logic is just to generate layer by skip of ten, but the last one has a safety check because it is not always necessary

**Comment by GuillaumeTh** [comment]
@frheault Can you correct the E128 pep8 errors please. After that, I think we are ready to merge

**Comment by MarioOcampo** [comment]
I did a test where I have:   `--multi_parameters 4`,   used 2 values of `tractogram_clustering_thr`, and (3 `bundle_pruning`  and 3 `model_clustering`) for all the bundles,   but for 1 bundle I have (1 `bundle_pruning`  and 1 `model_clustering` ).
At the beginning it loaded 4 configuration for all the bundles: (Parameters choice for bundle_A.trk, for the 4 executions are...)
When it loaded the one with (1 and 1) showed the error More multi-parameters executions than potential parameters
Then, the parameters choice is for 2 executions, for all the following bundles.

For the first ones it did 4  recobundlesx executions per tractogram_clustering_thr, and for the others 2.

The counter multi-parameters change if a bundle has less configurations.

Francois is fixing it (making it crash on purpose).

File: scilpy/segment/recobundlesx.py [hunk 20]
```diff
@@ -0,0 +1,300 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import logging
+from itertools import chain
```
Comment (jchoude) [inline scilpy/segment/recobundlesx.py]: Should be before logging.

File: scilpy/segment/recobundlesx.py [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+
+import numpy as np
```
Comment (jchoude) [inline scilpy/segment/recobundlesx.py]: Should be after dipy.

File: scripts/scil_recognize_single_bundle.py [hunk 22]
```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+    Compute a simple Recobundles (single-atlas & single-parameters).
+    The model need to be cleaned and lightweight.
```
Comment (jchoude) [inline scripts/scil_recognize_single_bundle.py]: need -> needs

File: scripts/scil_recognize_single_bundle.py [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+import nibabel as nib
+from nibabel.streamlines.array_sequence import ArraySequence
+import numpy as np
+
+from scilpy.io.utils import (add_overwrite_arg,
+                             assert_inputs_exist,
+                             assert_outputs_exists)
+
+
+def buildArgsParser():
```
Comment (jchoude) [inline scripts/scil_recognize_single_bundle.py]: _build_args_parser

File: scripts/scil_recognize_single_bundle.py [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+    p = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter,
+        description=__doc__,
+        epilog="Garyfallidis, E., Côté, M. A., Rheault, F., Sidhu, J., Hau, J.,"
+        "Petit, L., ... & Descoteaux, M. (2018). Recognition of white matter"
```
Comment (jchoude) [inline scripts/scil_recognize_single_bundle.py]: You just remove Fortin and Cunnane. You should either shorten it up more, or keep everyone.

Also, please add some line breaks. It doesn't look good when compared to the rest of the help message.

File: scripts/scil_recognize_multi_bundles.py [hunk 25]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+"""
+    Compute  RecobundlesX (multi-atlas & multi-parameters).
```
Comment (jchoude) [inline scripts/scil_recognize_multi_bundles.py]: One too many spaces between `Compute` and `RecobundlesX`.
